kernel/src/util/ring_buffer.rs

```diff
@@ -0,0 +1,569 @@
+        let next_index = (curr_index + len) % self.cap;
+        self.read_index.store(next_index, Ordering::Release);
+
+        self.len.fetch_sub(len, Ordering::Release);
+    }
+}
+
+impl RingBuffer<u8> {
+    /// Writes data from the [`VmReader`] to the [`RingBuffer`].
+    ///
+    /// Returns the number of bytes written.
+    pub fn write(&mut self, reader: &mut VmReader) -> usize {
+        let mut producer = Producer {
+            rb: self,
+            phantom: PhantomData,
+        };
+        producer.write(reader)
+    }
+
+    /// Reads data from the [`VmWriter`] to the [`RingBuffer`].
+    ///
+    /// Returns the number of bytes read.
+    pub fn read(&mut self, writer: &mut VmWriter) -> usize {
+        let mut consumer = Consumer {
+            rb: self,
+            phantom: PhantomData,
+        };
+        consumer.read(writer)
+    }
+}
+
+impl<T: Pod, R: Deref<Target = RingBuffer<T>>> Producer<T, R> {
+    const T_SIZE: usize = core::mem::size_of::<T>();
+
+    /// Pushes an item to the [`RingBuffer`].
+    ///
+    /// Returns an error if the ring buffer is full or
+    /// the underlying memory write failed.
+    pub fn push(&mut self, item: T) -> Result<()> {
+        let rb = &self.rb;
+        if rb.is_full() {
+            return_errno_with_message!(Errno::ENOMEM, "ring buffer is full")
+        }
+
+        let write_index = rb.write_index();
+        debug_assert!(write_index < rb.cap);
+
+        let segment_index = write_index * Self::T_SIZE;
+        let mut writer = rb.segment.writer().skip(segment_index).limit(Self::T_SIZE);
+        writer.write_val(&item)?;
+
+        rb.advance_write_index(write_index, 1);
+        Ok(())
+    }
+
+    /// Pushes a slice of items to the [`RingBuffer`].
+    ///
+    /// On success, all items are pushed to the ring buffer.
+    pub fn push_slice(&mut self, items: &[T]) -> Result<()> {
+        let nitems = items.len();
+        let rb = &self.rb;
+        let free_len = rb.free_len();
+        if free_len < nitems {
+            return_errno_with_message!(Errno::ENOMEM, "ring buffer is full or cannot fit all items")
+        }
+
+        let write_index = rb.write_index();
+        debug_assert!(write_index < rb.cap);
+        let segment_index = write_index * Self::T_SIZE;
+
+        if write_index + nitems > rb.cap {
+            // Write into two separate parts
+            rb.segment
+                .write_slice(segment_index, &items[..rb.cap - write_index])?;
+            rb.segment.write_slice(0, &items[rb.cap - write_index..])?;
+        } else {
+            rb.segment.write_slice(segment_index, items)?;
+        }
+
+        rb.advance_write_index(write_index, nitems);
+        Ok(())
+    }
+}
+
+impl<R: Deref<Target = RingBuffer<u8>>> Producer<u8, R> {
+    /// Writes data from the [`VmReader`] to the [`RingBuffer`].
+    ///
+    /// Returns the number of bytes written.
+    pub fn write(&mut self, reader: &mut VmReader) -> usize {
```

Comment:
This method should be called `write_fallable` and return `Result<usize, (Error, usize)>` so that the caller can distinguish between having no data/space and encountering page fault.